// webdriverio/build/commands/browser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.waitUntil = exports.url = exports.uploadFile = exports.touchAction = exports.switchWindow = exports.setWindowSize = exports.setTimeout = exports.setCookies = exports.saveScreenshot = exports.saveRecordingScreen = exports.reloadSession = exports.react$ = exports.react$$ = exports.pause = exports.newWindow = exports.keys = exports.getWindowSize = exports.getCookies = exports.executeAsync = exports.execute = exports.deleteCookies = exports.debug = exports.custom$ = exports.custom$$ = exports.call = exports.$ = exports.$$ = void 0;

const $$ = require('./browser/$$').default;

exports.$$ = $$;

const $ = require('./browser/$').default;

exports.$ = $;

const call = require('./browser/call').default;

exports.call = call;

const custom$$ = require('./browser/custom$$').default;

exports.custom$$ = custom$$;

const custom$ = require('./browser/custom$').default;

exports.custom$ = custom$;

const debug = require('./browser/debug').default;

exports.debug = debug;

const deleteCookies = require('./browser/deleteCookies').default;

exports.deleteCookies = deleteCookies;

const execute = require('./browser/execute').default;

exports.execute = execute;

const executeAsync = require('./browser/executeAsync').default;

exports.executeAsync = executeAsync;

const getCookies = require('./browser/getCookies').default;

exports.getCookies = getCookies;

const getWindowSize = require('./browser/getWindowSize').default;

exports.getWindowSize = getWindowSize;

const keys = require('./browser/keys').default;

exports.keys = keys;

const newWindow = require('./browser/newWindow').default;

exports.newWindow = newWindow;

const pause = require('./browser/pause').default;

exports.pause = pause;

const react$$ = require('./browser/react$$').default;

exports.react$$ = react$$;

const react$ = require('./browser/react$').default;

exports.react$ = react$;

const reloadSession = require('./browser/reloadSession').default;

exports.reloadSession = reloadSession;

const saveRecordingScreen = require('./browser/saveRecordingScreen').default;

exports.saveRecordingScreen = saveRecordingScreen;

const saveScreenshot = require('./browser/saveScreenshot').default;

exports.saveScreenshot = saveScreenshot;

const setCookies = require('./browser/setCookies').default;

exports.setCookies = setCookies;

const setTimeout = require('./browser/setTimeout').default;

exports.setTimeout = setTimeout;

const setWindowSize = require('./browser/setWindowSize').default;

exports.setWindowSize = setWindowSize;

const switchWindow = require('./browser/switchWindow').default;

exports.switchWindow = switchWindow;

const touchAction = require('./browser/touchAction').default;

exports.touchAction = touchAction;

const uploadFile = require('./browser/uploadFile').default;

exports.uploadFile = uploadFile;

const url = require('./browser/url').default;

exports.url = url;

const waitUntil = require('./browser/waitUntil').default;

exports.waitUntil = waitUntil;
