// webdriverio/build/commands/element.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.waitUntil = exports.waitForExist = exports.waitForEnabled = exports.waitForDisplayed = exports.waitForClickable = exports.touchAction = exports.shadow$ = exports.shadow$$ = exports.setValue = exports.selectByVisibleText = exports.selectByIndex = exports.selectByAttribute = exports.scrollIntoView = exports.saveScreenshot = exports.react$ = exports.react$$ = exports.moveTo = exports.isSelected = exports.isFocused = exports.isExisting = exports.isEnabled = exports.isDisplayedInViewport = exports.isDisplayed = exports.isClickable = exports.getValue = exports.getText = exports.getTagName = exports.getSize = exports.getProperty = exports.getLocation = exports.getHTML = exports.getCSSProperty = exports.getAttribute = exports.dragAndDrop = exports.doubleClick = exports.custom$ = exports.custom$$ = exports.click = exports.clearValue = exports.addValue = exports.$ = exports.$$ = void 0;

const $$ = require('./element/$$').default;

exports.$$ = $$;

const $ = require('./element/$').default;

exports.$ = $;

const addValue = require('./element/addValue').default;

exports.addValue = addValue;

const clearValue = require('./element/clearValue').default;

exports.clearValue = clearValue;

const click = require('./element/click').default;

exports.click = click;

const custom$$ = require('./element/custom$$').default;

exports.custom$$ = custom$$;

const custom$ = require('./element/custom$').default;

exports.custom$ = custom$;

const doubleClick = require('./element/doubleClick').default;

exports.doubleClick = doubleClick;

const dragAndDrop = require('./element/dragAndDrop').default;

exports.dragAndDrop = dragAndDrop;

const getAttribute = require('./element/getAttribute').default;

exports.getAttribute = getAttribute;

const getCSSProperty = require('./element/getCSSProperty').default;

exports.getCSSProperty = getCSSProperty;

const getHTML = require('./element/getHTML').default;

exports.getHTML = getHTML;

const getLocation = require('./element/getLocation').default;

exports.getLocation = getLocation;

const getProperty = require('./element/getProperty').default;

exports.getProperty = getProperty;

const getSize = require('./element/getSize').default;

exports.getSize = getSize;

const getTagName = require('./element/getTagName').default;

exports.getTagName = getTagName;

const getText = require('./element/getText').default;

exports.getText = getText;

const getValue = require('./element/getValue').default;

exports.getValue = getValue;

const isClickable = require('./element/isClickable').default;

exports.isClickable = isClickable;

const isDisplayed = require('./element/isDisplayed').default;

exports.isDisplayed = isDisplayed;

const isDisplayedInViewport = require('./element/isDisplayedInViewport').default;

exports.isDisplayedInViewport = isDisplayedInViewport;

const isEnabled = require('./element/isEnabled').default;

exports.isEnabled = isEnabled;

const isExisting = require('./element/isExisting').default;

exports.isExisting = isExisting;

const isFocused = require('./element/isFocused').default;

exports.isFocused = isFocused;

const isSelected = require('./element/isSelected').default;

exports.isSelected = isSelected;

const moveTo = require('./element/moveTo').default;

exports.moveTo = moveTo;

const react$$ = require('./element/react$$').default;

exports.react$$ = react$$;

const react$ = require('./element/react$').default;

exports.react$ = react$;

const saveScreenshot = require('./element/saveScreenshot').default;

exports.saveScreenshot = saveScreenshot;

const scrollIntoView = require('./element/scrollIntoView').default;

exports.scrollIntoView = scrollIntoView;

const selectByAttribute = require('./element/selectByAttribute').default;

exports.selectByAttribute = selectByAttribute;

const selectByIndex = require('./element/selectByIndex').default;

exports.selectByIndex = selectByIndex;

const selectByVisibleText = require('./element/selectByVisibleText').default;

exports.selectByVisibleText = selectByVisibleText;

const setValue = require('./element/setValue').default;

exports.setValue = setValue;

const shadow$$ = require('./element/shadow$$').default;

exports.shadow$$ = shadow$$;

const shadow$ = require('./element/shadow$').default;

exports.shadow$ = shadow$;

const touchAction = require('./element/touchAction').default;

exports.touchAction = touchAction;

const waitForClickable = require('./element/waitForClickable').default;

exports.waitForClickable = waitForClickable;

const waitForDisplayed = require('./element/waitForDisplayed').default;

exports.waitForDisplayed = waitForDisplayed;

const waitForEnabled = require('./element/waitForEnabled').default;

exports.waitForEnabled = waitForEnabled;

const waitForExist = require('./element/waitForExist').default;

exports.waitForExist = waitForExist;

const waitUntil = require('./element/waitUntil').default;

exports.waitUntil = waitUntil;
